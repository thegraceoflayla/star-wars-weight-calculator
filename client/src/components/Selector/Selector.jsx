import React from 'react';

const Selector = ({select, rotateDisplay, people, selectedPerson}) => (
  <div data-testid="selector" id="selector" className="col align-center m-4">
    <div id="people" className="row row-1 justify-content-center m-2">
      {
        people.map((person, i) => (
          <button data-testid="person" id={i} key={i} className={`btn ${person.name === selectedPerson ? 'btn-secondary' : 'btn-outline-secondary'} btn-lg m-1`} onClick={select}>{person.name}</button>
        ))
      }
    </div>
    <div id="scroll-buttons" className="row row-1 justify-content-center">
      <button id="left-button" className="btn btn-primary m-2 col-1" value="left" onClick={rotateDisplay}>
        <i className="fa-solid fa-chevron-left"></i>
      </button>
      <button id="right-button" className="btn btn-primary m-2 col-1" value="right" onClick={rotateDisplay}>
        <i className="fa-solid fa-chevron-right"></i>
      </button>
    </div>
  </div>
);

export default Selector;