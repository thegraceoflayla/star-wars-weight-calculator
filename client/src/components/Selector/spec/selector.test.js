import React from 'react';
import renderer from 'react-test-renderer';
import mockData from './mockData.js'
import Selector from '../Selector.jsx';

it('renders once props are recieved', (done) => {
  const selector = renderer.create(<Selector people={mockData} selectedPerson={'Luke Skywalker'} select={() => {}} rotateDisplay={() => {}} />);
  expect(selector).toMatchSnapshot();
  done();
});

it('displays people array prop as buttons', (done) => {
  const selector = renderer.create(<Selector people={mockData} selectedPerson={'Luke Skywalker'} select={() => {}} rotateDisplay={() => {}} />);
  let buttons = selector.toJSON().children[0].children;
  expect(buttons.every(button => button.type === 'button')).toBe(true);
  expect(buttons.every((button, i) => button.children[0] === mockData[i].name)).toBe(true);
  done();
});

it('emphasizes the currently selected person', (done) => {
  const selector = renderer.create(<Selector people={mockData} selectedPerson={'Luke Skywalker'} select={() => {}} rotateDisplay={() => {}} />);
  let buttons = selector.toJSON().children[0].children;
  let unselectedClass = 'btn-outline-secondary';
  let selectedClass = 'btn-secondary';
  let unselectedButtons = buttons.slice(1);
  expect(unselectedButtons.every(button => button.props.className.includes(unselectedClass))).toBe(true);
  expect(buttons[0].props.className.includes(selectedClass)).toBe(true);
  done();
});

it('calls select function when button is clicked', (done) => {
  let numberOfSelectCalls = 0;
  const selector = renderer.create(<Selector people={mockData} selectedPerson={'Luke Skywalker'} select={() => { numberOfSelectCalls++; }} rotateDisplay={() => {}} />);
  let buttons = selector.toJSON().children[0].children;
  buttons.forEach((button) => button.props.onClick());
  expect(numberOfSelectCalls).toBe(5);
  done();
});

it('calls rotateDisplay function when arrow button is clicked', (done) => {
  let rotateDisplayCalled = false;
  const selector = renderer.create(<Selector people={mockData} selectedPerson={'Luke Skywalker'} select={() => {}} rotateDisplay={() => { rotateDisplayCalled = true; }} />);
  expect(rotateDisplayCalled).toBe(false);
  selector.toJSON().children[1].children[0].props.onClick();
  expect(rotateDisplayCalled).toBe(true);
  done();
});