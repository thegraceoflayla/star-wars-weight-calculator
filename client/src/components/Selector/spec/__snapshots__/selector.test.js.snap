// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders once props are recieved 1`] = `
<div
  className="col align-center m-4"
  data-testid="selector"
  id="selector"
>
  <div
    className="row row-1 justify-content-center m-2"
    id="people"
  >
    <button
      className="btn btn-secondary btn-lg m-1"
      data-testid="person"
      id={0}
      onClick={[Function]}
    >
      Luke Skywalker
    </button>
    <button
      className="btn btn-outline-secondary btn-lg m-1"
      data-testid="person"
      id={1}
      onClick={[Function]}
    >
      C-3PO
    </button>
    <button
      className="btn btn-outline-secondary btn-lg m-1"
      data-testid="person"
      id={2}
      onClick={[Function]}
    >
      R2-D2
    </button>
    <button
      className="btn btn-outline-secondary btn-lg m-1"
      data-testid="person"
      id={3}
      onClick={[Function]}
    >
      Darth Vader
    </button>
    <button
      className="btn btn-outline-secondary btn-lg m-1"
      data-testid="person"
      id={4}
      onClick={[Function]}
    >
      Leia Organa
    </button>
  </div>
  <div
    className="row row-1 justify-content-center"
    id="scroll-buttons"
  >
    <button
      className="btn btn-primary m-2 col-1"
      id="left-button"
      onClick={[Function]}
      value="left"
    >
      <i
        className="fa-solid fa-chevron-left"
      />
    </button>
    <button
      className="btn btn-primary m-2 col-1"
      id="right-button"
      onClick={[Function]}
      value="right"
    >
      <i
        className="fa-solid fa-chevron-right"
      />
    </button>
  </div>
</div>
`;
